File: kabanbot/services/llm.py
from litellm import acompletion
import logging
from typing import List, Dict, Any
from kabanbot.config import settings
from kabanbot.prompts import SYSTEM_PROMPT

logger = logging.getLogger(__name__)


class LLMService:
    def __init__(self):
        self.system_prompt = SYSTEM_PROMPT

    async def summarize(self, messages: List[Dict[str, Any]]) -> str:
        """
        Summarizes a list of messages using the configured LLM.

        Args:
            messages: List of dicts with keys 'username', 'text', 'timestamp'.

        Returns:
            The summary string.
        """
        if not messages:
            return "No messages to summarize."

        def format_message(msg: Dict[str, Any]) -> str:
            username = msg["username"]
            text = msg["text"]
            reply_to_text = msg.get("reply_to_text")
            reply_to_username = msg.get("reply_to_username")

            if reply_to_text and reply_to_username:
                return f'{username} (replying to {reply_to_username}: "{reply_to_text}"): {text}'
            return f"{username}: {text}"

        conversation_text = "\n".join(format_message(msg) for msg in messages)

        try:
            response = await acompletion(
                model=settings.LLM_MODEL,
                api_key=settings.LLM_API_KEY,
                base_url=settings.LLM_BASE_URL,
                messages=[
                    {"role": "system", "content": self.system_prompt},
                    {
                        "role": "user",
                        "content": f"Messages:\n```\n{conversation_text}\n```",
                    },
                ],
            )
            return response.choices[0].message.content
        except Exception as e:
            logger.error(f"Error calling LLM: {e}")
            return "Sorry, I couldn't generate a summary at this time."

File: services/llm.go
package services

import (
	"context"
	"fmt"

	"github.com/sashabaranov/go-openai"
	"kabanbot/config"
)

// LLMService manages interactions with the configured LLM provider.
type LLMService struct {
	client *openai.Client
	model  string
}

// NewLLMService instantiates an LLMService configured with settings from config.
func NewLLMService(cfg *config.Config) *LLMService {
	openaiCfg := openai.DefaultConfig(cfg.LLMAPIKey)
	if cfg.LLMBaseURL != "" {
		openaiCfg.BaseURL = cfg.LLMBaseURL
	}

	client := openai.NewClientWithConfig(openaiCfg)
	return &LLMService{
		client: client,
		model:  cfg.LLMModel,
	}
}

// GenerateGreeting generates a dynamic greeting using the LLM.
func (s *LLMService) GenerateGreeting(ctx context.Context, firstName string) (string, error) {
	resp, err := s.client.CreateChatCompletion(
		ctx,
		openai.ChatCompletionRequest{
			Model: s.model,
			Messages: []openai.ChatCompletionMessage{
				{
					Role:    openai.ChatMessageRoleSystem,
					Content: "You are a helpful and friendly assistant. Generate a warm, welcoming, and slightly playful greeting message for a Telegram user in response to their /start command. Address them by their first name.",
				},
				{
					Role:    openai.ChatMessageRoleUser,
					Content: fmt.Sprintf("My name is %s. Start the conversation!", firstName),
				},
			},
		},
	)
	if err != nil {
		return "", err
	}

	if len(resp.Choices) == 0 {
		return "", fmt.Errorf("no choices returned from LLM")
	}

	return resp.Choices[0].Message.Content, nil
}
